### ISSUE TITLE
cd_adf: badly formatted example issue while running on cluster

### ISSUE THREAD

**Issue by musram** [VowpalWabbit/vowpal_wabbit#3657]
We are facing "'VW::vw_exception'
  what():  cb_adf: badly formatted example, only one cost can be known.", when the code is executed on the cluster. But on stand-alone, it works fine.

The training data which we have passed is in a format specified by the VW and it works on a stand-alone machine. So no issue with the data.

We changed the version to 8.11.0 but still, the same error persists. So we can rule out the version issues.

The command is
./vw.binary -P 4113882 \
--hash all \
--data data.csv \
--cache_file temp.cache \
--final_regressor cb.vwmodel \
--bit_precision 24 \
-k \
--passes 1 \
--cb_explore_adf \
--cb_type ips \
--save_resume \
--holdout_off \
--span_server 10.39.32.28 \
--total 1 \
--node 000000 \
--unique_id 1050

Are there any other params that we need to pass in the cluster environment?

**Comment by jackgerrits** [comment]
This kind of bug is notoriously hard to find. I could be wrong but in the past things of this sort have been due to the way VW reuses example objects and is essentially dependent on the order of examples received.

Could the reason be why it doesn't repro locally be because VW sees a different ordering of data? 

If it is possible to upload the datafile I can try different ordering of inputs to try and repro?

**Comment by musram** [comment]
Hi, @jackgerrits  here is the file of sample data that is causing the above error. We have checked it from our side. Could you check it?
[part-00000-3e7310a3-a894-419a-8926-561435caf8b9-c000.txt](https://github.com/VowpalWabbit/vowpal_wabbit/files/7968131/part-00000-3e7310a3-a894-419a-8926-561435caf8b9-c000.txt)

**Comment by musram** [comment]
No there is no error if not in a cluster environment.  And even in a cluster if the error pops up, we shuffle the data, and then it works without error.

**Comment by jackgerrits** [comment]
I've been trying random shuffles of the data locally to try and get lucky and repro the issue but no luck so far.

**Comment by jackgerrits** [comment]
This issue can happen in many different ways and we've been fixed several occurrences of this class of bug over time so it's hard to pin it down. We have some plans that should resolve this kind of bug at a more fundamental level.
I'm going to close this issue though as I can't get a repro from this dataset. If you can find a repro please let me know!

**Comment by musram** [comment]
Thanks, @jackgerrits  for your effort. The same dataset which we shared with you gives the mentioned error in a cluster but not in a normal environment.
As I said we are solving this by reshuffling the data. Hoping for a permanent fix for this problem by Vowpal wabbit.

**Comment by jackgerrits** [comment]
Could you do me a favour and check if adding `--onethread` in the cluster environment removes the issue?

**Comment by JohnLangford** [comment]
Another question is: can we reproduce by just using the same bad shard at each node?  (For two processes on the same node?)

**Comment by musram** [comment]
@jackgerrits,  We tried with --onethread but have the same problem.

Running VW command ... ./vw.binary -P 4113882
        --hash all
        --data data.dat
        --cache_file temp.cache
        --final_regressor cb.vwmodel
        --bit_precision 28
        -k
        --passes 1
        --onethread
        --cb_explore_adf
        --cb_type ips
        --save_resume
        --holdout_off
        --span_server 10.39.32.28
        --total 1
        --node 000000
        --unique_id 0119

final_regressor = cb.vwmodel
Num weight bits = 28
learning rate = 0.5
initial_t = 0
power_t = 0.5
creating cache_file = temp.cache
Reading datafile = data.dat
num sources = 1
Enabled reductions: gd, scorer, csoaa_ldf, cb_adf, cb_explore_adf_greedy, shared_feature_merger
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
vw example #32(cb_adf.cc:275): cb_adf: badly formatted example, only one cost can be known.
terminate called after throwing an instance of 'VW::vw_exception'
  what():  cb_adf: badly formatted example, only one cost can be known.

**Comment by musram** [comment]
@JohnLangford we ran the same data file on two different nodes but it failed.

Running VW command ... ./vw.binary -P 4113882
        --hash all
        --data data.dat
        --cache_file temp.cache
        --final_regressor cb.vwmodel
        --bit_precision 28
        -k
        --passes 1
        --onethread
        --cb_explore_adf
        --cb_type ips
        --save_resume
        --holdout_off
        --span_server 10.39.32.28
        --total 2
        --node 000000
        --unique_id 0135
final_regressor = cb.vwmodel
Num weight bits = 28
learning rate = 0.5
initial_t = 0
power_t = 0.5
creating cache_file = temp.cache
Reading datafile = data.dat
num sources = 1
Enabled reductions: gd, scorer, csoaa_ldf, cb_adf, cb_explore_adf_greedy, shared_feature_merger
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
vw example #32(cb_adf.cc:275): cb_adf: badly formatted example, only one cost can be known.
terminate called after throwing an instance of 'VW::vw_exception'
  what():  cb_adf: badly formatted example, only one cost can be known.

Regarding the shard, the data which we shared with you is too small to be shard. And also as I said shuffling the data or any change to the data resolves the issue. Change here is like deleting a few lines etc.

**Comment by JohnLangford** [comment]
If the two nodes are two different processes on the same machine, does it still fail?

**Comment by musram** [comment]
@JohnLangford  yes it fails when the two nodes are two different processes on the same machine.

**Comment by jackgerrits** [comment]
I had a go running this locally but wasn't able to reproduce. Can you describe how you ran it locally?

**Comment by musram** [comment]
@jackgerrits we used both command lines using the VW binary and also checked using python. Locally it works but it fails on a cluster.  If we shuffle the data it works. As VW is not supported for the distributed environments we use a map reduced with VW binary.

**Comment by jackgerrits** [comment]
Can you please describe precisely how you ran it locally (full command lines etc) because I ran it map reduced locally and didn't hit the bug. So I am trying to determine what I did differently to you

**Comment by musram** [comment]
@jackgerrits  sorry for the delay. I will let you know all the information you need from my side in a couple of days.

**Comment by musram** [comment]
Hi @jackgerrits . Is it possible to have a meeting with you on a video call? It will be easy for both of us.

**Comment by jackgerrits** [comment]
Apologies for the delay this got away from me. We run an open meeting each fortnight, so today and in another fortnight. Perhaps you could join? If not then we can try and sort something out. [How to join](https://github.com/VowpalWabbit/agenda/issues/21)

**Comment by musram** [comment]
@jackgerrits  my team member will join the meeting.

**Comment by musram** [comment]
@jackgerrits  thanks. Whatever solution you provided is working.

**Comment by musram** [comment]
@jackgerrits  . 

We made the changes as suggested by you. Today we had the same  "[critical] vw (cb_adf.cc:286): cb_adf: badly formatted example, only one cost can be known but found 6. Example number=0, tag=F9=2"

This is the first example. We ran the pipeline again for a couple of times and then there was no error.

So I think still the problem persists.

**Comment by jackgerrits** [comment]
Can you create a local repro? Without that it is very hard to try and fix this

**Comment by jackgerrits** [comment]
@musram if a local repro is not possible then as a test to see if the issue is caused by what I suspect would you be able to test out your scenario using these changes? https://github.com/VowpalWabbit/vowpal_wabbit/pull/3774 As in build this branch and run your scenario with the resulting binary

**Comment by musram** [comment]
@jackgerrits we will try tomorrow and let you know the results.

**Comment by jackgerrits** [comment]
Can you please provide a local repro? I cannot look into this without it. Having them join the meeting will not really be helpful if we cannot debug

**Comment by jackgerrits** [comment]
Please feel free to reopen if you can work out a repro.
